### PR TITLE
opensearch: init at 2.5.0

### PR DESCRIPTION
**Pull request by shyim**
###### Description of changes

I maintain the PHP client for Opensearch and packaged it myself as I tested the newest release. Also relates to packaging request #151464

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [X] aarch64-linux
  - [X] x86_64-darwin
  - [X] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/search/opensearch/default.nix
Comment (Kranzes) [inline pkgs/servers/search/opensearch/default.nix:18]:
```suggestion
stdenv.mkDerivation (rec {
  pname = "opensearch";
  version = "2.1.0";

  src = fetchurl {
    url = "https://artifacts.opensearch.org/releases/bundle/opensearch/${version}/opensearch-${version}-linux-x64.tar.gz";
    sha256 = "sha256-vM1zfg9ORVTTkY8BGh43nr2p5TRzxwtCZKooznGWEeM=";
  };
```

Comment (Kranzes) [reply pkgs/servers/search/opensearch/default.nix:18]: Updated to latest version.

File: pkgs/servers/search/opensearch/default.nix
Comment (ajs124) [inline pkgs/servers/search/opensearch/default.nix:41]: since this fetches a `linux-x64.tar.gz`, isn't this architecture specific?

Comment (shyim) [reply pkgs/servers/search/opensearch/default.nix:41]: I am working on macbook.. Deleting the jre in zip and using own works good :)

**Comment by shyim** [comment]
@Kranzes I adopted the most, but I would not restrict the packaging only for the official provided packages. Officially, is only Linux amd64/arm64 packaged. The difference is the packaged Java. 

As we use our own Java, we can support any platform our own java supports 😅

**Review by drupol — APPROVED** [review]
LGTM

**Comment by fballiano** [comment]
it seems one of the workflow hung and won't run, can we do something about it?

**Comment by drupol** [comment]
@GrahamcOfBorg build opensearch
@GrahamcOfBorg test opensearch

**Comment by drupol** [comment]
Maybe @SuperSandro2000 may help in here to push the frustration away.

**Comment by mweinelt** [comment]
@ofborg eval

File: pkgs/servers/search/opensearch/default.nix
Comment (mweinelt) [inline pkgs/servers/search/opensearch/default.nix:45]: Can we get a homepage link in here? Ideally a link to their source code repository, where we can submit issues and browse the code. Should be https://github.com/opensearch-project/OpenSearch.

File: pkgs/servers/search/opensearch/default.nix
Comment (mweinelt) [inline pkgs/servers/search/opensearch/default.nix:19]:
```suggestion
    hash = "";
```

Can we migrate this to an SRI hash instead? If you build this using an empty hash you should get one.

File: pkgs/servers/search/opensearch/default.nix
Comment (mweinelt) [inline pkgs/servers/search/opensearch/default.nix:18]:
```suggestion
  version = "2.3.0";

  src = fetchurl {
    url = "https://artifacts.opensearch.org/releases/bundle/opensearch/${version}/opensearch-${version}-linux-x64.tar.gz";
```

Tie the source URL to the version, so we don't have to modify it on every update. That being said, the latest release seems to be 2.5.0.

And what about other targets than x86_64-linux, are there any?

File: pkgs/servers/search/opensearch/default.nix
Comment (fballiano) [inline pkgs/servers/search/opensearch/default.nix:41]:
```suggestion
    description = "Open Source, Distributed, RESTful Search Engine";
    homepage = "https://github.com/opensearch-project/OpenSearch";
```

File: pkgs/servers/search/opensearch/default.nix
Comment (shyim) [reply pkgs/servers/search/opensearch/default.nix:18]: We use the JDK from nix. So it makes no difference for any platform. I have it running on MacOS :)

**Comment by shyim** [comment]
Applied all comments and tested locally on my Mac
image

File: pkgs/servers/search/opensearch/default.nix
Comment (drupol) [inline pkgs/servers/search/opensearch/default.nix:18]:
```suggestion
    url = "https://artifacts.opensearch.org/releases/bundle/opensearch/${version}/opensearch-${version}-linux-x64.tar.gz";
```

Comment (mweinelt) [reply pkgs/servers/search/opensearch/default.nix:18]: That one still applies :)

Comment (shyim) [reply pkgs/servers/search/opensearch/default.nix]: Added it :)

File: pkgs/servers/search/opensearch/default.nix
Comment (fballiano) [inline pkgs/servers/search/opensearch/default.nix:18]:
```suggestion
    url = "https://artifacts.opensearch.org/releases/bundle/opensearch/${version}/opensearch-${version}-linux-x64.tar.gz";
```

Comment (shyim) [reply pkgs/servers/search/opensearch/default.nix]: haha did it now too 😂

File: pkgs/servers/search/opensearch/default.nix
Comment (mweinelt) [inline pkgs/servers/search/opensearch/default.nix:38]:
```suggestion
  installPhase = ''
    runHook preInstall  

    mkdir -p $out
    cp -R bin config lib modules plugins $out

    substituteInPlace $out/bin/opensearch \
      --replace 'bin/opensearch-keystore' "$out/bin/opensearch-keystore"

    wrapProgram $out/bin/opensearch \
      --prefix PATH : "${lib.makeBinPath [ util-linux gnugrep coreutils ]}" \
      --set JAVA_HOME "${jre_headless}"

    wrapProgram $out/bin/opensearch-plugin --set JAVA_HOME "${jre_headless}"

    runHook preInstall
  '';
```

And let's follow the best practice here, to make overriding more predictable.

Comment (shyim) [reply pkgs/servers/search/opensearch/default.nix]: done :)

File: pkgs/servers/search/opensearch/default.nix
Comment (mweinelt) [inline pkgs/servers/search/opensearch/default.nix:41]:
```suggestion
    runHook postInstall
```

That should be the last fixup for now.

Comment (drupol) [inline pkgs/servers/search/opensearch/default.nix:41]:
```suggestion
    runHook postInstall
```

Comment (shyim) [reply pkgs/servers/search/opensearch/default.nix]: done :)

File: pkgs/servers/search/opensearch/opensearch-home-fix.patch
Comment (mweinelt) [inline pkgs/servers/search/opensearch/opensearch-home-fix.patch:21]:
```
❯ OPENSEARCH_HOME=/tmp/opensearch ./result/bin/opensearch-node
could not find java in bundled jdk at /tmp/opensearch/jdk/bin/java
```

So what it actually requires is essentially like a `JAVA_HOME`? But I cannot just pass in for example `openjdk.out`, because it has `bin/java`, but not `jdk/bin/java`.

```
❯ OPENSEARCH_HOME=/nix/store/ch095in4hr481iqv11s55ks9ksl2c49i-openjdk-19.0.1+10/ ./result/bin/opensearch-node
could not find java in bundled jdk at /nix/store/ch095in4hr481iqv11s55ks9ksl2c49i-openjdk-19.0.1+10//jdk/bin/java
```

So then I symlink myself something together, and now I'm stuck here:

````
❯ OPENSEARCH_HOME=/tmp/opensearch ./result/bin/opensearch
Error: Could not find or load main class org.opensearch.tools.java_version_checker.JavaVersionChecker
Caused by: java.lang.ClassNotFoundException: org.opensearch.tools.java_version_checker.JavaVersionChecker
````

What does it take to get this running? I think we really need a small module to encapsulate this requirement.

Comment (shyim) [reply pkgs/servers/search/opensearch/opensearch-home-fix.patch:21]: see this as example how to get it up running :)
https://github.com/cachix/devenv/blob/main/src/modules/services/elasticsearch.nix#L35-L80

Comment (mweinelt) [reply pkgs/servers/search/opensearch/opensearch-home-fix.patch:21]: If they haven't diverge enough already, a compatible module would be amazing.

Comment (shyim) [reply pkgs/servers/search/opensearch/opensearch-home-fix.patch:21]: Make the existing elastic search one compatible or a new one?

Comment (mweinelt) [reply pkgs/servers/search/opensearch/opensearch-home-fix.patch:21]: That really depends on how compatible they still are.

Comment (shyim) [reply pkgs/servers/search/opensearch/opensearch-home-fix.patch:21]: I added a module `services.opensearch.enable = true` and works

image

Comment (mweinelt) [reply pkgs/servers/search/opensearch/opensearch-home-fix.patch:21]: Cool, what do you think about adding a small nixos test to continuously validate that everything is still working?

Comment (shyim) [reply pkgs/servers/search/opensearch/opensearch-home-fix.patch:21]: I will look tomorrow at the moment I don't have a machine with kvm enabled

Comment (shyim) [reply pkgs/servers/search/opensearch/opensearch-home-fix.patch:21]: test is added :)

File: nixos/modules/module-list.nix
Comment (mweinelt) [inline nixos/modules/module-list.nix:1046]:
```suggestion
  ./services/search/meilisearch.nix
  ./services/search/opensearch.nix
```

Proper ordering.

File: nixos/modules/services/search/opensearch.nix
Comment (mweinelt) [inline nixos/modules/services/search/opensearch.nix:24]: Instead of build the config by hand, try `pkgs.formats.yaml` from RFC 42.

https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md

File: nixos/modules/services/search/opensearch.nix
Comment (mweinelt) [inline nixos/modules/services/search/opensearch.nix:72]: Put those below `services.opensearch.settings`, so you can later use `format.generate "opensearch.yml cfg.settings`. The options should then use the proper path that the config expects.

**Review by mweinelt — CHANGES_REQUESTED** [review]
Thanks, that looks pretty good already. Two issues:

- The module should use the `settings` option and the formatter from RFC42
- The tests should be in its own commit `nixos/tests/opensearch: init`.

Check the botamusique module for an example on how those bits work.

File: nixos/modules/services/search/opensearch.nix
Comment (shyim) [reply nixos/modules/services/search/opensearch.nix:72]: So we leave cluster name and so. Just set inside enable them into `settings` and use `settings` to generate that file

Comment (mweinelt) [reply nixos/modules/services/search/opensearch.nix:72]: Whatever goes into the config files should be below `settings`.

Comment (shyim) [reply nixos/modules/services/search/opensearch.nix:72]: Alright. I will look after Fosdem :)

Comment (drupol) [reply nixos/modules/services/search/opensearch.nix:72]: Hey you're going to Fosdem ! I'll be there too :) I'll be sticking the Nix room on Sunday afternoon hit me up !

Comment (shyim) [reply nixos/modules/services/search/opensearch.nix]: @mweinelt done :)

File: nixos/tests/opensearch.nix
Comment (SuperSandro2000) [inline nixos/tests/opensearch.nix:18]:
```suggestion
    machine.start()
    machine.wait_for_unit("opensearch.service")
    machine.wait_for_open_port(9200)

    machine.succeed(
        "curl --fail localhost:9200"
    )
  '';
```

File: nixos/modules/services/search/opensearch.nix
Comment (SuperSandro2000) [inline nixos/modules/services/search/opensearch.nix:23]:
```suggestion
```
Those are legacy comments which get copy pasted. We can remove them.

File: nixos/modules/services/search/opensearch.nix
Comment (SuperSandro2000) [inline nixos/modules/services/search/opensearch.nix:29]: This should use mkEnableOption

File: nixos/modules/services/search/opensearch.nix
Comment (SuperSandro2000) [inline nixos/modules/services/search/opensearch.nix:36]: This should be using mkPackageOption

File: nixos/modules/services/search/opensearch.nix
Comment (SuperSandro2000) [inline nixos/modules/services/search/opensearch.nix:141]: nit

```suggestion
```

File: nixos/modules/services/search/opensearch.nix
Comment (SuperSandro2000) [inline nixos/modules/services/search/opensearch.nix:171]:
```suggestion
      preStart = optionalString (!config.boot.isContainer) ''
        # Only set vm.max_map_count if lower than ES required minimum
        # This avoids conflict if configured via boot.kernel.sysctl
        if [ $(${pkgs.procps}/bin/sysctl -n vm.max_map_count) -lt 262144 ]; then
          ${pkgs.procps}/bin/sysctl -w vm.max_map_count=262144
        fi
     '' + ''
```
we don't need ot inline this here. Also $() is usually used instead of backticks.

File: nixos/modules/services/search/opensearch.nix
Comment (SuperSandro2000) [inline nixos/modules/services/search/opensearch.nix:180]: Normally we use tmpfiles for directory creation

File: nixos/modules/services/search/opensearch.nix
Comment (SuperSandro2000) [inline nixos/modules/services/search/opensearch.nix:176]: For symlinks we can also use tmpfiles

File: nixos/modules/services/search/opensearch.nix
Comment (SuperSandro2000) [inline nixos/modules/services/search/opensearch.nix:188]:
```suggestion
```
looks like debugging code to me. This could also leak secrets into logs.

File: nixos/modules/services/search/opensearch.nix
Comment (SuperSandro2000) [inline nixos/modules/services/search/opensearch.nix:196]:
```suggestion
        sed -e '#logs/gc.log#${cfg.dataDir}/logs/gc.log#' -i ${configDir}/jvm.options \
```
no need to draw in an extra dependency

File: nixos/modules/services/search/opensearch.nix
Comment (SuperSandro2000) [inline nixos/modules/services/search/opensearch.nix:198]: preStart is always run as root

File: pkgs/servers/search/opensearch/default.nix
Comment (SuperSandro2000) [inline pkgs/servers/search/opensearch/default.nix:36]: We don't really need to wrap gnugrep and coreutils since they are always there on nixos.

**Review by SuperSandro2000 — COMMENTED** [review]
just some minor thongs from me. Overhaul it looks already pretty good.

File: nixos/modules/services/search/opensearch.nix
Comment (shyim) [reply nixos/modules/services/search/opensearch.nix:176]: Do you have an example for me? like other service as reference what you mean :)

File: nixos/modules/services/search/opensearch.nix
Comment (mweinelt) [reply nixos/modules/services/search/opensearch.nix:198]: Is it? Do you have a reference for that?

File: nixos/modules/services/search/opensearch.nix
Comment (mweinelt) [reply nixos/modules/services/search/opensearch.nix:180]: I wouldn't insist on using tmpfiles, they're a bit on the painful end of things.

File: nixos/modules/services/search/opensearch.nix
Comment (shyim) [reply nixos/modules/services/search/opensearch.nix:188]: AHH yes. It was for debugging. I removed it

File: pkgs/servers/search/opensearch/default.nix
Comment (shyim) [reply pkgs/servers/search/opensearch/default.nix:36]: I use that package also on macOs. I guess it makes sense to leave it still here or? 🤔

File: nixos/modules/services/search/opensearch.nix
Comment (shyim) [reply nixos/modules/services/search/opensearch.nix]: Seems so. I have tested it and adjusted it

File: nixos/modules/services/search/opensearch.nix
Comment (SuperSandro2000) [reply nixos/modules/services/search/opensearch.nix:163]: https://github.com/NixOS/nixpkgs/blob/35d1efaf52d8e3956a9a91ed5972fbc2ad1c1f48/nixos/modules/services/cluster/hadoop/yarn.nix#L144-L146

also see ``man tmpfiles.d`` for all the details

File: nixos/modules/services/search/opensearch.nix
Comment (SuperSandro2000) [reply nixos/modules/services/search/opensearch.nix]: I might have mixed up ExecStartPre with preStart from our start script.

Also that is not right by default. I think the fact I based this knowledge on was using PermissionsStartOnly=true which is off by default. Sorry about that.

File: nixos/modules/services/search/opensearch.nix
Comment (talyz) [reply nixos/modules/services/search/opensearch.nix:163]: tmpfiles should not be used for this! It's much better to do it as it's already done, as that makes sure the symlinks are set up when the service starts. Doing the same thing with tmpfiles can result in the service being run before the symlinks are created.

**Comment by shyim** [comment]
@mweinelt like so? copied the title like from https://github.com/NixOS/nixpkgs/commit/fe34d10e57b76696738b3a4a4352480d4e0ffd2f 😅

File: pkgs/servers/search/opensearch/default.nix
Comment (mweinelt) [inline pkgs/servers/search/opensearch/default.nix:43]:
```suggestion
  passthru.tests = {
    inherit (nixosTests) opensearch;
  }:
```

This links the package to the nixos test, which will be run by ofborg when the package gets updated. Should go into the `nixos/tests/opensearch` commit.

Comment (shyim) [reply pkgs/servers/search/opensearch/default.nix:44]: done :)

**Review by mweinelt — APPROVED** [review]

**Comment by mweinelt** [comment]
Waiting for a second approval. Thanks for sinking your time into this.

**Comment by mweinelt** [comment]
Oh, no! I forgot: New modules also require a changelog entry. See `nixos/doc/manual/release-notes/rl-2305.section.md` and the relevant checkboxes in the top post.

**Review by fballiano — APPROVED** [review]

**Comment by shyim** [comment]
Updated release page, hope fully done. 😅

**Review by mweinelt — APPROVED** [review]

**Review by drupol — COMMENTED** [review]
Massive work, thanks!

**Review by talyz — CHANGES_REQUESTED** [review]
Looks good overall! However, I thought it would be nice to get rid of `PermissionsStartOnly` and have `DynamicUser` and `StateDirectory` setup working, so I've opened https://github.com/shyim/nixpkgs/pull/1. It also includes some other minor fixes to the documentation and the start script.

**Comment by shyim** [comment]
@talyz it's merged. many many thanks! ❤️ .

Do I need to cleanup the commits or it's fine so?

File: nixos/modules/services/search/opensearch.nix
Comment (talyz) [inline nixos/modules/services/search/opensearch.nix:26]: Oops, I broke the manual build with this. Sorry about that! Should be
```suggestion
    enable = mkEnableOption (lib.mdDoc "OpenSearch");
```

**Comment by talyz** [comment]
No problem! I think it's fine to keep them as separate commits.

File: nixos/modules/services/search/opensearch.nix
Comment (shyim) [reply nixos/modules/services/search/opensearch.nix]: looks now good :)

**Review by talyz — APPROVED** [review]

**Comment by talyz** [comment]
Great work! Thanks!

**Comment by drupol** [comment]
Massive!

**Pull request merged**
